descartes_planner/include/descartes_planner/ladder_graph.h

```diff
@@ -0,0 +1,224 @@
+  }
+
+  /**
+   * @brief isLast tests to see if a given index is the last one in the graph
+   */
+  bool isLast(size_type index) const noexcept
+  {
+    return index + 1 == size();
+  }
+
+  /**
+   * @brief isFirst tests to see if given index is the first in the graph
+   */
+  bool isFirst(size_type index) const noexcept
+  {
+    return index == 0;
+  }
+
+  /**
+   * @brief vertex returns a pointer to the data that constitutes the Nth vertex in the Jth row
+   *        where N = index & J = rung
+   */
+  const double* vertex(size_type rung, size_type index) const
+  {
+    return getRung(rung).data.data() + (dof_ * index);
+  }
+
+  /**
+   * @brief The number of rungs
+   */
+  size_type size() const noexcept
+  {
+    return rungs_.size();
+  }
+
+  size_type dof() const noexcept
+  {
+    return dof_;
+  }
+
+  /**
+   * @brief assign Consumes the given edge list and assigns it to the rung-index given by 'rung'
+   */
+  void assignEdges(size_type rung, std::vector<EdgeList>&& edges) // noexcept?
+  {
+    getEdges(rung) = std::move(edges);
+  }
+
+  /**
+   * @brief assignRung Special helper function to assign a solution set associated with a Descartes point &
+   *        it's meta-info. Also resizes the associated edge list to the size of 'sols'.
+   * @param sols All of the joint solutions for this point.
+   */
+  void assignRung(size_type index, descartes_core::TrajectoryID id, descartes_core::TimingConstraint time,
+                  const std::vector<std::vector<double>>& sols)
+  {
+    Rung& r = getRung(index);
+    r.id = id;
+    r.timing = time;
+    r.data.reserve(sols.size() * dof_);
+    for (const auto& sol : sols)
+    {
+      r.data.insert(r.data.end(), sol.cbegin(), sol.cend());
+    }
+    // Given this new vertex set, build an edge list for each
+    getEdges(index).resize(r.data.size());
+  }
+
+  void removeVertices(size_type index)
+  {
+    rungs_.erase(std::next(rungs_.begin(), index));
+  }
+
+  void removeEdges(size_type index)
+  {
+    edges_.erase(std::next(edges_.begin(), index));
+  }
+
+  void removeRung(size_type index)
+  {
+    removeVertices(index);
+    removeEdges(index);
+  }
+
+  void clearVertices(size_type index)
+  {
+    rungs_[index].data.clear();
+  }
+
+  void clearEdges(size_type index)
+  {
+    edges_[index].clear();
+  }
+
+  /**
+   * @brief insertRung Adds a new rung at the 'index'-th position. E.g., insertRung(0) will add a new
+   *        rung to the beginning of the graph and the previous 0th index is now at 1.
+   */
+  void insertRung(size_type index)
+  {
+    rungs_.insert(std::next(rungs_.begin(), index), Rung() );
+    edges_.insert(std::next(edges_.begin(), index), std::vector<EdgeList>() );
+  }
+
+  /**
+   * @brief Clears all existing rungs & associated data
+   */
+  void clear()
+  {
+    rungs_.clear();
+    edges_.clear();
+  }
+
+private:
+  const size_type dof_;
+  std::vector<Rung> rungs_;
```

Comment:
Not a big deal, but it really feels like combining rungs and edges into their own class would have been cleaner (perhaps at the expense of efficiency).